lara: adopt back gun in TR1

lahm86 opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This is a draft PR to allow for some testing to begin. I have replicated TR2's rifle module in TR1 effectively, we can merge this later if everything goes well (so ignore the messy commit for the time being 😄).

The injection means we can adopt TR2's approach of delegating the animation to the object itself. The mesh is separated from Lara's torso so LMK if this is positioned suitably on her back. And tests around firing and general behaviour would be appreciated. Saving and loading too, although see the note below about legacy.

TR1 also had a "Fix shotgun targeting" option which I'm not sure if we still need. Does TR2 have that issue?

Remaining TODOs:
- [x] Handle legacy BSON and ATI saves (these won't work currently unless Lara is unarmed or doesn't have the shotgun)
- [x] Update gym gun injections
- [x] Merge `gun_rifle.c`
- [x] Remove legacy shotgun targeting option
- [x] Handle carrying rifles between levels (temporarily disabled to avoid crash)
- [x] Document for builders/create a suitable wad export

Ref: https://github.com/LostArtefacts/TRXInjectionTool/pull/145

lahm86 committed:
lara: adopt back gun in TR1

rr- commented:
>TR1 also had a "Fix shotgun targeting" option which I'm not sure if we still need. Does TR2 have that issue?

IIRC this was authored by @walkawayy – pinging for attention :relieved: 

This is the original issue: https://github.com/LostArtefacts/TRX/issues/51
Original diff: 20c2d81f811e104c465b3a18dda72004e66f7c90

aredfan commented:
Yes that's correct, the shotgun targeting issue is a TR1-specific bug, it doesn't affect OG TR2 or TR2X.  I've also confirmed the bug is fixed in the PR build with said option disabled.

It's probably a coincidence but I noticed the texture issue with Lara's glove whilst holding a shotgun, is now fixed in this PR's build. https://github.com/LostArtefacts/TRX/discussions/2582#discussioncomment-12389508 , for reference.

**Edit:** Nvm I just noticed your other PR. Great job with fixing the glove issue, and thank you. 👍

walkawayy commented:
> > TR1 also had a "Fix shotgun targeting" option which I'm not sure if we still need. Does TR2 have that issue?
> 
> IIRC this was authored by @walkawayy – pinging for attention 😌
> 
> This is the original issue: #51 Original diff: [20c2d81](https://github.com/LostArtefacts/TRX/commit/20c2d81f811e104c465b3a18dda72004e66f7c90)

Basically in OG TR1, the shotgun has a faster fire rate when Lara isn't aiming at a target. She also keeps shooting even when the target moves out of sight (like TR4/5 targeting) though that is at odds with all other TR1-3 guns where Lara stops shooting in that scenario.

TR2 shotgun seems like it behaves properly, so I would say it's safe to port TR2 shotgun behavior to TR1. It is technically different behavior than the OG TR1 which is why we made it optional originally. I'm in favor of reducing the mass amount of options by making undesirable bugs fixed by default, but I totally get wanting to maintain OG behavior even if it is slightly buggy. I would be interested in what @lahm86 and @aredfan think.

lahm86 commented:
Thanks, @aredfan. I forgot to mention the glove fix 😄 
That makes sense @walkawayy, thanks for clarifying. I suppose with the TR4 "no lock" option that we now have, the OG behaviour of losing the target could more or less be simulated. I'd be in favour of removing the specific shotgun option as the buggy fast firing I don't think serves any use - Lara would switch to normal rate when locking on to something else I believe?

I'll continue with the rest of this task in the meantime. Thanks again.

aredfan commented:
> TR2 shotgun seems like it behaves properly, so I would say it's safe to port TR2 shotgun behavior to TR1. It is technically different behavior than the OG TR1 which is why we made it optional originally. I'm in favor of reducing the mass amount of options by making undesirable bugs fixed by default, but I totally get wanting to maintain OG behavior even if it is slightly buggy. I would be interested in what @lahm86 and @aredfan think.

Agree with all your points @walkawayy.  I remember the shotgun targeting fix was one of the options introduced in the early days of this project, which was a good thing for visibility.

I would say it's okay to retire the option, my main reasoning is because of the inconsistent behaviour between the shotgun and other weapons, something that was rectified in TR2+.

lahm86 committed:
gun/rifle: merge rifle control in TRX

lahm86 committed:
config: remove TR1 enhanced shotgun option

The enhanced shotgun targeting option is redundant with the TR2
controlled animations; the weapon lock mode can be set to no lock to
achieve a similar result to OG TR1.

lahm86 committed:
tr1/savegame: handle legacy back gun conversion

This allows legacy saves to switch to the back gun concept on load.

lahm86 committed:
lara: make gun status carry over optional

This updates TR1 to make the feature where Lara starts a new level
with her guns drawn tied to the same option that remembers her last
gun type. This also now extends to TR2, and specific handling is
applied for the rifle type weapons.

lahm86 commented:
This should be ready for review now.

- The mesh has been adjusted slightly to better match the OG position on her back
- Injection updated for the gym
- For legacy saves, because we don't have a reference to the matching animation number for the gun object, Lara will "snap" into the relevant state on load. So for example, if you have an old save and she's half-way through drawing the shotgun, it will load with the shotgun fully drawn. Similar for undrawing. LMK if this is ok. New saves from here on out will store all the relevant information, per TR2.
- Remembering guns between levels (OG TR1 quirk) required handling for the back gun as it needs the item to have been created during the draw animation. This feature now exists in TR2, but in both games it has been combined with remembering the gun type. A particular area to check would be going from the likes of Diving Area into 40 Fathoms, with various different guns equipped. In TR1 you can disable the `remove_weapons` gf sequence in Natla's Mines to test similar there.

aredfan commented:
- [x] If Lara has the grenade launcher drawn, proceeding to the next level causes her to put the gun away, and take it out again

> 1. `/play 5`
> 2. use all items cheat
> 3. draw the grenade launcher
> 4. slow down the game's speed
> 5. use the level skip cheat and observe Lara's behaviour

**Edit:** It'll be worth checking Lara's behaviour with the grenade launcher when proceeding from DA to 40F.

lahm86 commented:
Thanks, @aredfan - should be fixed. I also fixed the back gun remaining reflective if you fly cheat off the Midas hand (and a related bug where the braid remained reflective).

aredfan commented:
> Thanks, @aredfan - should be fixed. I also fixed the back gun remaining reflective if you fly cheat off the Midas hand (and a related bug where the braid remained reflective).

The gun grenade issue is resolved, thank you. 👍 

I've ran into a few more issues:

- [x] In develop, if Lara has a shotgun on her back and she's turning to gold. The shotgun disappears when Lara's backpack turns to gold.  In the PR build, the shotgun persists.

> <details><summary>Develop</summary>
> 
> ![20251027_222422_Palace_Midas](https://github.com/user-attachments/assets/f6dc2046-c558-4f87-8187-e79bd97cd7ff)
> 
> </details>
> 
> <details><summary>PR</summary>
> 
> ![20251027_222348_Palace_Midas](https://github.com/user-attachments/assets/5d1a24a8-745d-4873-9a7a-6d8a805ac50b)
> 
> </details>

- [x] If Lara has the shotgun on her back when she meets Bacon, then Bacon has a shotgun visible.

> <details><summary>Screenshot</summary>
> 
> ![20251027_221847_Atlantis](https://github.com/user-attachments/assets/7ca3e300-edb3-487b-97ef-5106b9d0ff46)
> 
> </details>

- [ ] ~~In develop, if Lara is eaten by the T-Rex, her shotgun disappears immediately.  This doesn't happen in the PR build.~~  

> Warning: graphic images.
> 
> <details><summary>Develop</summary>
> 
> ![20251027_222731_Lost_Valley](https://github.com/user-attachments/assets/6b820e56-8b91-4941-b99b-1cb577201a86)
> 
> </details>
> 
> <details><summary>PR</summary>
> 
> ![20251027_222635_Lost_Valley](https://github.com/user-attachments/assets/9c4187bb-a735-46d2-ad3d-d5e65937e0bd)
> 
> </details>

**Edit:** Perhaps the last issue can be considered a bugfix.

lahm86 commented:
Thanks, I'll fix bacon Lara tomorrow. I'm sort of on the fence about gold Lara, with reflections on it looks quite good but off it does look jarring. Maybe we can just hide it in this case. And I agree t-rex Lara feels more like a bug fix.

lahm86 commented:
@aredfan - updated. The gun will disappear on Midas' hand, and from Bacon Lara, but will remain when killed by t-rex.

aredfan reviewed (approved):
Thank you @lahm86, the issues are resolved. 👍 

I agree that it'd be cool to have Lara's shotgun turn to gold.  Perhaps we could revisit the feasibility of this in the future.
